### PR TITLE
test OAuth requests using pact.io as mock server OP#40804

### PR DESCRIPTION
**Pull request by individual-it**
using the framework of pact.io to mock the OAuth requests of OpenProject and testing the `request` function

I'm open for discussion if that is the framework we want to use to mock the requests. The advantage is that, we can simulate real requests and responses without having to spin up a complete NC or OpenProject server.
The framework writes the requests and response into a JSON file and in future we can use that to verify that what we expect from OpenProject is really what they send.

here a short intro in how it works: https://docs.pact.io/getting_started/how_pact_works#consumer-testing

### CODE REVIEW

**Comment by wielinde** [comment]
Hi everybody,

I am not familiar with the PHP toolchain etc. So I cannot be of great help here. At OP we use similar stuff, for instance [_Puffing Billy_](https://github.com/oesmith/puffing-billy), which allows for mocking services on the browser level. It definitely has its legitimacy.

Needless to say, that it will not save us from:
- Manual tests
- Real integration tests
which we need to tackle and setup one by one, too.

File: tests/lib/Service/OpenProjectAPIServiceTest.php [hunk 1]
```diff
@@ -39,8 +102,83 @@ public function urlsDataProvider(): array {
...
+	public function testValidateOpenProjectURL(string $url, bool $expected)	{
 		$result = OpenProjectAPIService::validateOpenProjectURL($url);
 		$this->assertSame($expected, $result);
 	}
+
+	public function testRequestUsingOAuthToken() {
+		$consumerRequest = new ConsumerRequest();
+		$consumerRequest
+			->setMethod('GET')
+			->setPath('/api/v3/work_packages')
```
Comment (kiranparajuli589) [inline tests/lib/Service/OpenProjectAPIServiceTest.php:114]: using a variable for this path will be useful since it is used multiple times below.

**Review by eneiluj — APPROVED** [review]
I'm fine with any choice you make for testing.

File: tests/lib/Service/OpenProjectAPIServiceTest.php (hunk 1 above)
Comment (individual-it) [reply tests/lib/Service/OpenProjectAPIServiceTest.php]: done, moved the path and also the URL of the mock-server into a variable, please review again

**Review by kiranparajuli589 — COMMENTED** [review]
looks good :+1:

**Review by kiranparajuli589 — APPROVED** [review]

**Pull request merged**